relay-server/src/actors/upstream.rs

```diff
@@ -466,29 +475,39 @@ impl Handler<Authenticate> for UpstreamRelay {
             self.config.upstream_descriptor()
         );
 
-        self.auth_state = AuthState::RegisterRequestChallenge;
         let request = RegisterRequest::new(&credentials.id, &credentials.public_key);
 
         let future = self
             .send_query(request)
             .into_actor(self)
             .and_then(|challenge, slf, ctx| {
                 log::debug!("got register challenge (token = {})", challenge.token());
-                slf.auth_state = AuthState::RegisterChallengeResponse;
                 let challenge_response = challenge.create_response();
 
                 log::debug!("sending register challenge response");
                 let fut = slf.send_query(challenge_response).into_actor(slf);
                 ctx.notify(PumpHttpMessageQueue);
                 fut
             })
-            .map(|_, slf, _ctx| {
+            .map(|_, slf, ctx| {
                 log::info!("relay successfully registered with upstream");
                 slf.auth_state = AuthState::Registered;
+                slf.backoff.reset();
+                slf.last_authentication = Some(Instant::now());
+                ctx.notify_later(Authenticate, slf.reauthentication_interval);
             })
             .map_err(|err, slf, ctx| {
                 log::error!("authentication encountered error: {}", LogError(&err));
-                slf.auth_state = AuthState::Error;
+
+                if let Some(last_authentication) = slf.last_authentication {
+                    if last_authentication + slf.max_reauthentication_interval < Instant::now() {
+                        // we are past our grace period we can no longer claim we are authenticated
+                        slf.auth_state = AuthState::Error;
+                    }
+                } else {
+                    // we never managed to authenticate and now we just got an error
+                    slf.auth_state = AuthState::Error;
```

Comment:
Shouldn't we retry in the error case as well? Otherwise, a Relay will stay in error and never try to reauthenticate. Likewise, if we're not over the grace period, nothing will check again.

Reply:
I'll have to check exactly but my impression/intention was that the normal retry-with-backoff will happen in this case.
Will double check.

Reply:
You are right, the regular retry kicks in after that.